Code/GraphMol/Subset.cpp

```diff
@@ -0,0 +1,346 @@
+//
+//  Copyright (C) 2025 Hussein Faara, Brian Kelley and other RDKit Contributors
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+#include <GraphMol/RDKitBase.h>
+#include <GraphMol/GraphMol.h>
+#include <GraphMol/QueryAtom.h>
+#include <GraphMol/QueryBond.h>
+#include "Subset.h"
+
+namespace RDKit {
+namespace {
+
+inline void copyComputedProps(const ROMol &src, ROMol &dst) {
+  dst.updateProps(src);
+  for(auto &v : dst.getPropList(true, false)) {
+    if(v != RDKit::detail::computedPropName)
+      dst.clearProp(v);
+  }
+}
+
+static void copySelectedAtomsAndBonds(::RDKit::RWMol &extracted_mol,
+                                      const RDKit::ROMol &reference_mol,
+                                      SubsetInfo &selection_info,
+				      const SubsetOptions &options) {
+  auto &[selectedAtoms, selectedBonds, atomMapping, bondMapping] =
+      selection_info;
+  for (const auto &ref_atom : reference_mol.atoms()) {
+    if (!selectedAtoms[ref_atom->getIdx()]) {
+      continue;
+    }
+
+    std::unique_ptr<::RDKit::Atom> extracted_atom{
+      options.copyAsQuery ? new QueryAtom(*ref_atom) : ref_atom->copy() };
+
+    static constinit bool updateLabel = false;
+    static constinit bool takeOwnership = true;
+    atomMapping[ref_atom->getIdx()] = extracted_mol.addAtom(
+        extracted_atom.release(), updateLabel, takeOwnership);
+  }
+
+  for (const auto &ref_bond : reference_mol.bonds()) {
+    if (!selectedBonds[ref_bond->getIdx()]) {
+      continue;
+    }
+
+    std::unique_ptr<::RDKit::Bond> extracted_bond{
+      options.copyAsQuery ? new QueryBond(*ref_bond) : ref_bond->copy() };
+
+    // Check the stereo atoms
+    auto &atoms = extracted_bond->getStereoAtoms();
+    if(atoms.size() == 2) {
+      auto map1 = atomMapping.find(atoms[0]);
+      auto map2 = atomMapping.find(atoms[1]);
+      if(map1 != atomMapping.end() && map2 != atomMapping.end()) {
+	atoms[0] = map1->second;
+	atoms[1] = map2->second;
+      } else {
+	atoms.clear(); // We couldn't map the stereo atoms
+      }
+    }
+    
+    for(auto &atomidx : atoms) {
+      auto map = atomMapping.find(atomidx);
+      if(map != atomMapping.end())
+	atomidx = map->second;
+    }
+    
+    extracted_bond->setBeginAtomIdx(atomMapping[ref_bond->getBeginAtomIdx()]);
+    extracted_bond->setEndAtomIdx(atomMapping[ref_bond->getEndAtomIdx()]);
+
+    static constinit bool takeOwnership = true;
+    auto num_bonds =
+        extracted_mol.addBond(extracted_bond.release(), takeOwnership);
+    bondMapping[ref_bond->getIdx()] = num_bonds - 1;
+
+    // we need to update rings now
+    if(reference_mol.getRingInfo()->isInitialized()) {
+      extracted_mol.getRingInfo()->reset();
+    }
+  }
```

Comment:
Since we create the molecule just before the call to this function, we're never going to need to do this. Still, I guess it doesn't hurt to include it.

However, I don't think we need to do this for every bond that we add. Can't we move it outside of the loop and do:
```suggestion
    // we need to update rings now
    if(selectedBonds.any() && reference_mol.getRingInfo()->isInitialized()) {
      extracted_mol.getRingInfo()->reset();
    }
  }
```

Reply:
Resolved